tests/test_virtualcenter.py

```diff
@@ -0,0 +1,146 @@
+
+
+def test_content(provider):
+    """Test that a content node is retrieved"""
+    content = provider.content
+
+    assert content is not None
+    provider._service_instance.RetrieveContent.assert_called_with()
+
+
+def test_version(provider):
+    """Test that the version is returned correctly"""
+    provider.content.about.version = '6.5'
+    ver = provider.version
+
+    assert ver == '6.5'
+
+
+def test_get_obj_list(provider, mocker):
+    """Test that we get the right objects"""
+    mocked_folder = mocker.MagicMock()
+    mocked_container = mocker.MagicMock(view=[1, 2, 3])
+    provider.content.viewManager.CreateContainerView.return_value = mocked_container
+    object_list = provider._get_obj_list(vim.VirtualMachine, mocked_folder)
+
+    provider.content.viewManager.CreateContainerView.assert_called_with(
+        mocked_folder, [vim.VirtualMachine], True)
+    assert object_list == [1, 2, 3]
+
+
+def test_get_obj(provider, mocker):
+    """Test that _get_obj() does the right calls"""
+    mock_obj = mocker.MagicMock()
+    mock_obj.name = 'mock'
+    mocked_get_list = mocker.MagicMock(return_value=[mocker.MagicMock(name='first'), mock_obj])
+    mocker.patch.object(provider, '_get_obj_list', mocked_get_list)
+    obj = provider._get_obj(vim.VirtualMachine, 'mock')
+
+    assert obj is mock_obj
+    mocked_get_list.assert_called_with(vim.VirtualMachine, None)
+
+
+def test_get_obj_doesnt_exist(provider, mocker):
+    """Test that _get_obj() returns None when it can't find an object"""
+    mocked_get_list = mocker.MagicMock(return_value=[mocker.MagicMock(name='first')])
+    mocker.patch.object(provider, '_get_obj_list', mocked_get_list)
+    obj = provider._get_obj(vim.VirtualMachine, 'mock')
+
+    assert obj is None
+    mocked_get_list.assert_called_with(vim.VirtualMachine, None)
+
+
+def test_build_filter_spec(provider):
+    """Test that the _build_filter_spec() method correctly builds a filter spec"""
+    start_entity = vim.VirtualMachine('start_entity')
+    property_spec = vmodl.query.PropertyCollector.PropertySpec()
+    property_spec.all = True
+    filter_spec = provider._build_filter_spec(start_entity, property_spec)
+
+    assert isinstance(filter_spec, vmodl.query.PropertyCollector.FilterSpec)
+    assert filter_spec.propSet == [property_spec]
+    assert len(filter_spec.objectSet) == 1
+    assert isinstance(filter_spec.objectSet[0], vmodl.query.PropertyCollector.ObjectSpec)
+    assert filter_spec.objectSet[0].obj is start_entity
+    assert len(filter_spec.objectSet[0].selectSet) == 9
+    assert filter_spec.objectSet[0].selectSet[0].name == 'resource_pool_traversal_spec'
+    assert filter_spec.objectSet[0].selectSet[0].type == vim.ResourcePool
+    assert filter_spec.objectSet[0].selectSet[0].path == 'resourcePool'
+    assert len(filter_spec.objectSet[0].selectSet[0].selectSet) == 2
+    assert filter_spec.objectSet[0].selectSet[0].selectSet[0].name == 'resource_pool_traversal_spec'
```

Comment:
**OPTIONAL** those lines have a massive amount of duplication, this certainly can be expressed more nicely